src/Uno.UI/UI/Xaml/Controls/Image/ImageSizeHelper.cs

```diff
@@ -1,122 +1,159 @@
 ﻿using System;
-using System.Collections.Generic;
-using System.Text;
 using Windows.Foundation;
 using Windows.UI.Xaml.Media;
+using Uno.UI;
 using static Windows.UI.Xaml.Media.Stretch;
 
 namespace Windows.UI.Xaml.Controls
 {
 	internal static class ImageSizeHelper
 	{
-		public static void MeasureSource(this Image image, Windows.Foundation.Rect parent, ref Windows.Foundation.Rect child)
+		public static void MeasureSource(this Image image, Size finalSize, ref Rect child)
 		{
 			switch (image.Stretch)
 			{
 				case UniformToFill:
-					var uniformToFillScale = (child.Width * parent.Height >= child.Height * parent.Width)
-						? parent.Height / child.Height // child is flatter than parent
-						: parent.Width / child.Width; // child is taller than parent
+					var uniformToFillScale = (child.Width * finalSize.Height >= child.Height * finalSize.Width)
+						? finalSize.Height / child.Height // child is flatter than parent
+						: finalSize.Width / child.Width; // child is taller than parent
 					child.Width *= uniformToFillScale;
 					child.Height *= uniformToFillScale;
 					break;
 
 				case Uniform:
-					var uniformScale = (child.Width * parent.Height > child.Height * parent.Width)
-						? parent.Width / child.Width // child is taller than parent
-						: parent.Height / child.Height; // child is flatter than parent
+					var uniformScale = (child.Width * finalSize.Height > child.Height * finalSize.Width)
+						? finalSize.Width / child.Width // child is taller than parent
+						: finalSize.Height / child.Height; // child is flatter than parent
 					child.Width *= uniformScale;
 					child.Height *= uniformScale;
 					break;
 
 				case Fill:
-					child.Width = parent.Width;
-					child.Height = parent.Height;
+					child.Width = finalSize.Width;
+					child.Height = finalSize.Height;
 					break;
 
 				case None:
 					break;
 			}
 		}
 
-		public static void ArrangeSource(this Image image, Windows.Foundation.Rect parent, ref Windows.Foundation.Rect child)
+		public static void ArrangeSource(this Image image, Size finalSize, ref Rect child)
 		{
+			var stretch = image.Stretch;
 			// In order to match UWP behaviors, in some specific cases the image must be left align
-			var isForcedLeft = (image.Stretch == None || image.Stretch == UniformToFill) && parent.Width <= child.Width;
+			//var isForcedLeft = (stretch == None || stretch == UniformToFill) && finalSize.Width <= child.Width;
+			var isForcedLeft = false;
+
+			double GetPositionForDimension(bool isForcedToZero, double imageDimension, double calculatedPosition)
```

Comment:
Not sure to see the point to pout this nested to the `ArrangeSource` method (I don't remember is the compilator will be able to put it as static or if it will make some closure). 

Anyway a negative impact is that you are always computing the `calculatedPosition` even if you don't use it